### ISSUE TITLE
How can I read the value of the Buffer?

### ISSUE THREAD

**Issue by shamoons** [googleapis/nodejs-speech#399]
I have a response like:
```
{ name: '7373656846233364184',
  metadata:
   { type_url:
      'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
     value:
      <Buffer 08 64 12 0c 08 aa 8d b2 e9 05 10 c8 a5 fc b9 02 1a 0c 08 f3 8d b2 e9 05 10 b0 f1 aa b1 02> },
  done: true,
  response:
   { type_url:
      'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeResponse',
     value:
      <Buffer 12 f3 01 0a e7 01 0a 27 46 6f 72 20 53 65 70 74 65 6d 62 65 72 20 31 73 74 2e 20 48 6f 77 20 64 6f 20 79 6f 75 20 72 65 67 69 73 74 65 72 3f 15 bb d9 ... > },
  result: 'response' }
```
I tried:
```
                const metadata = JSON.parse(getResponse.metadata.value.toString());
                const response = JSON.parse(getResponse.response.value.toString());
```
But that has an error:

```
Unhandled rejection SyntaxError: Unexpected token in JSON at position 0
```

**Comment by bcoe** [comment]
@shamoons what API method are you using, could you share a snippet of code?

**Comment by shamoons** [comment]
```
                return speechClient.operationsClient.getOperation({ name })
```

**Comment by bcoe** [comment]
@shamoons what do you get if you run:

```
getResponse.response.value.toString('utf8')
```

**Comment by shamoons** [comment]
```

��������
       �������
```

**Comment by bcoe** [comment]
@alexander-fenster this one seems a bit odd. Reading the jsdoc, it seems that the response for long running operations should be an object, @shamoons is getting raw binary data.

**Comment by alexander-fenster** [comment]
@shamoons The contents of the operation is serialized protobuf. You can decode it using functions from `protobufjs` directly, but there is an easier way.

Assuming you have a `speech` client somewhere nearby (e.g. `const speech = new Speech.SpeechClient()`), do this:

```js
const decodedOperation = new gax.Operation(getResponse, speech._descriptors.longrunning.longRunningRecognize);
console.log('decoded result:', decodedOperation.result);
console.log('decoded metadata:', decodedOperation.metadata);
```
This is just a little bit of a grey zone (undocumented private member access) but it will save you some time. If you wish, go with `protobufjs` directly instead.

Having that said - I'd like to know your use case. If you use `longRunningRecognize` directly from a speech client, you already receive an instance of `gax.Operation` with all fields decoded:

```js
  const [operation] = await speech.longRunningRecognize(request);
  // operation is gax.Operation: http://googleapis.github.io/gax-nodejs/classes/Operation.html
  operation.promise().then(([result, metadata, rawOperation]) => {
    console.log('result:', result);
    console.log('metadata:', metadata);
    console.log('raw operation:', rawOperation); // this is similar to your getResponse
  });
```

The only use case I can think of is that you're getting the raw operation from a client running somewhere else and you need to read the contents of this operation. In that case, use the code from my first snippet.

**Comment by alexander-fenster** [comment]
TL;DR of my previous comment: you shouldn't need to access `operationsClient` directly unless you need to access operations started elsewhere. In most use cases, it's just this:

```js
const [operation] = await speech.longRunningRecognize(request);
const [result, metadata] = await operation.promise();
```

**Comment by shamoons** [comment]
I actually am accessing it elsewhere. I’m triggering the job with a lambda and can’t necessarily wait. I then, async, check the status periodically

**Comment by alexander-fenster** [comment]
@shamoons Yes this use case makes sense. My first snippet (the one that creates a `gax.Operation` based on your `getResponse`) is a possible solution.

Alternative way of doing that would be to use `protobufjs` directly by loading proto files from `node_modules/` and decoding those messages manually. As a background, both metadata and response are serialized instances of `google.protobuf.Any` which, unfortunately, is not supported by `protobufjs` directly, but can be decoded using the following code I just wrote for you:

```js
const protobufjs = require('protobufjs');

// load proto files once. Note v1 below: if you use speech.v1p1beta1, use it in the path
const root = protobuf.loadSync([ // note: synchronous file read - use .load() to use callback API
  './node_modules/@google-cloud/speech/protos/google/cloud/speech/v1/cloud_speech.proto',
  './node_modules/google-gax/protos/google/protobuf/timestamp.proto'
]);

// helper function to decode google.protobuf.Any
function decodeProtobufAny(protobufAny) {
  const typeName = protobufAny.type_url.replace(/^.*\//, '');
  const type = root.lookupType(typeName);
  return type.decode(protobufAny.value);
}
```

With this helper function (and a loaded `root`), decoding is easy:

```js
const decodedMetadata = decodeProtobufAny(getResponse.metadata);
const decodedResponse = decodeProtobufAny(getResponse.response);
```

**Comment by shamoons** [comment]
Awesome. Thank you! For the time being, I'm doing:
```
                return axios.request({
                    method: 'get',
                    url: `https://speech.googleapis.com/v1p1beta1/operations/${googleName}?key=${process.env.GOOGLE_SPEECH_API_KEY}`,
                })
``` 

as I'm a bit uncomfortable with undocumented methods. But I hope that others may benefit from this. Thanks for your incredible work!

**Comment by shamoons** [comment]
Also, @alexander-fenster , I had posted this [Q on SO](https://stackoverflow.com/questions/57044074/with-google-cloud-speech-to-text-and-the-node-js-sdk-how-can-i-read-the-value-o/57058604#57058604) as well. If you'd like imaginary internet points, just write an answer and I'll delete mine and accept yours.

**Comment by bcoe** [comment]
@shamoons I'm going to go ahead and close this, as I think Alex's recommendations should do the trick for you :+1: 

@alexander-fenster thoughts on anywhere we might document this centrally, since this "one neat trick" can work on any gRPC client ... perhaps `google-cloud-node` could use a `docs/` folder.

**Comment by alexander-fenster** [comment]
@shamoons Thanks, I'm too lazy to copy my response over there for imaginary internet points (even though I sometimes do collect them :) )

As for something being documented, the `protobufjs` solution does not use any undocumented methods and leaves you within gRPC HTTP/2 space (no `axios`, no HTTP/1 connection). But your solution is totally fine as well.

**Comment by alexander-fenster** [comment]
@bcoe re: documenting: since I now have a good solution for this, I'd better work on the generator side to make it possible to create an instance of `gax.Operation` with all types (metadata and response) defined from within a client. It's a common use case for long running APIs running in cloud functions or Lambda and we need to have a good support for it.

**Comment by repentsinner** [comment]
Here's a more complete example of @alexander-fenster's LRO decoder - `Operation()` required a `backoffSetting` to compile for me.

```typescript
import * as gax from 'google-gax'
// there might be a better way to find backoffSettings, autocomplete suggested this
import { createDefaultBackoffSettings } from 'google-gax/build/src/gax';
// pick your preferred pinned speech version
const speech = require('@google-cloud/speech').v1p1beta1

const speechClient = new speech.SpeechClient()

async function checkLROProgress(name:string) {
  const [operation] = await speechClient.operationsClient.getOperation({name})
  const decodedOperations = new gax.Operation(operation, speechClient._descriptors.longrunning.longRunningRecognize, createDefaultBackoffSettings())
  
  console.log(decodedOperations.name)
  console.log(decodedOperations.done)
  console.log(decodedOperations.result)
  console.log(decodedOperations.metadata)
}

void checkLROProgress('{your LRO id here}')
```

**Comment by idudinov** [comment]
hey guys! looks like the recent move to Typescript has changed things a bit.
SpeechClient doesn't expose `operationsClient`, neither public or private. Can somebody please show an example with new client library?

LROs still look unusable in cases when environment (like Firebase functions) won't be able to wait for the completion too long.

**Comment by willbattel** [comment]
Running into the same inconvenience with the AutoML Vision node library. [The docs](https://cloud.google.com/vision/automl/docs/create-datasets#get-operation) cite the `operationsClient.getOperation` method that upon my testing returns an unusable buffer in the response field unless I then decode it. This means I can't get back my newly created dataset or model IDs unless I insert some form of decoding, as you've shown in this thread. That behavior doesn't seem optimal.

I don't know if/where the Speech/Vision/AutoML teams overlap here but, if there is any connection, I'd really appreciate some improved docs for an officially-recommended decoding method or, even better, for the response to be decoded before being handed to us.

**Comment by alexander-fenster** [comment]
@willbattel Thank you! I agree this needs to be improved. I filed https://github.com/googleapis/gapic-generator-typescript/issues/170 for tracking purposes (we are converting all the libraries to TypeScript and this will only be taken care of in the TypeScript generator; AutoML will also be converted).

For the time being, feel free to use the decoding code from this thread. When we implement a good API for that, it will essentially do the same.

**Comment by feimosi** [comment]
I'm having a related issue but in my case the response is simply empty...

It happens only when I run it with Google Cloud URI. If I use it with a local file then it works fine (as long as the file is under 1-minute long). I'm running the beta API (`v1p1beta1)` because of the MP3s support.

Anyone else experiences this as well?

Here's how I run it:
```ts
const [operation] = await this.client.longRunningRecognize(request);
const [response, metadata, runningOperation] = await operation.promise();

console.log(response);
console.log(metadata);
console.log(runningOperation);
```

Prints (note `results: []`):
```sh
LongRunningRecognizeResponse { results: [] }
LongRunningRecognizeMetadata {
  progressPercent: 100,
  startTime: Timestamp {
    seconds: Long { low: 1603475277, high: 0, unsigned: false },
    nanos: 715931000
  },
  lastUpdateTime: Timestamp {
    seconds: Long { low: 1603475310, high: 0, unsigned: false },
    nanos: 152903000
  },
  uri: 'gs://test/audio/ea8d1100-32dd-49f3-aae2-564abe106651.mp3'
}
{
  name: '5537819703969830356',
  metadata: {
    type_url: 'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeMetadata',
    value: <Buffer 08 64 12 0c 08 cd ae cc fc 05 10 f8 fa b0 d5 02 1a 0b 08 ee ae cc fc 05 10 d8 ba f4 48 22 41 67 73 3a 2f 2f 62 69 74 63 61 73 74 2d 63 6c 69 70 73 2f ... 46 more bytes>
  },
  done: true,
  response: {
    type_url: 'type.googleapis.com/google.cloud.speech.v1p1beta1.LongRunningRecognizeResponse',
    value: <Buffer >
  },
  result: 'response'
}
```

**Comment by alexander-fenster** [comment]
@feimosi You're getting empty response, which in most cases means that some of the request parameters were incorrect (i.e. did not match the audio, or the audio was considered silent). You are getting `done = true` (operation is completed) but empty response (so it did not detect any speech in your MP3 file).

If you publish this MP3 somewhere for us, along with the request parameters you use, we can help figure out the right parameters, or ask the API team to investigate.
